Provisioning: Add validation for Job specifications

MissingRoberto opened the pull request:
## What this PR does / why we need it:

Adds comprehensive validation for Job specifications in the provisioning API. Previously, Jobs had no validation (marked with FIXME), which could lead to invalid job configurations being accepted.

## Changes:
- **Added job validation** in new \`apps/provisioning/pkg/jobs\` package
- **Validates all job actions**: pull, push, pr, migrate, delete, move
- **Action-specific validation**:
  - Push jobs: validates branch names and paths
  - Delete jobs: validates paths and resource references
  - Move jobs: validates source/target paths and resource references
- **Leverages existing utilities**:
  - Uses \`git.IsValidGitBranchName()\` for branch validation
  - Uses \`safepath.IsSafe()\` for path validation (prevents path traversal, invalid characters, etc.)
- **Returns error directly** instead of field.ErrorList (cleaner API)
- **Called from admission controller** in register.go

# Issues

Fixes https://github.com/grafana/git-ui-sync-project/issues/373

MissingRoberto committed:
Validate Job Specs

MissingRoberto committed:
Add comprehensive unit test coverage for job validator

- Added 8 new test cases to improve coverage from 88.9% to ~100%
- Tests for migrate action without options
- Tests for delete/move actions with resources (missing kind)
- Tests for move action with valid resources
- Tests for move/delete with both paths and resources
- Tests for move action with invalid source paths
- Tests for push action with valid paths

Now covers all validation paths including resource validation and
edge cases for all job action types.

MissingRoberto committed:
Add integration tests for job validation

Added comprehensive integration tests that verify the job validator properly
rejects invalid job specifications via the API:

- Test job without action (required field)
- Test job with invalid action
- Test pull job without pull options
- Test push job without push options
- Test push job with invalid branch name (consecutive dots)
- Test push job with path traversal attempt
- Test delete job without paths or resources
- Test delete job with invalid path (path traversal)
- Test move job without target path
- Test move job without paths or resources
- Test move job with invalid target path (path traversal)
- Test migrate job without migrate options
- Test valid pull job to ensure validation doesn't block legitimate requests

These tests verify that the admission controller properly validates job specs
before they are persisted, ensuring security (path traversal prevention) and
data integrity (required fields/options).

MissingRoberto committed:
Remove valid job test case from integration tests

Removed the positive test case as it's not necessary for validation testing.
The integration tests now focus solely on verifying that invalid job specs
are properly rejected by the admission controller.

MissingRoberto committed:
Fix movejob_test to expect validation error at creation time

Updated the 'move without target path' test to expect the job creation
to fail with a validation error, rather than expecting the job to be
created and then fail during execution.

This aligns with the new job validation logic which rejects invalid
job specs at the API admission control level (422 Unprocessable Entity)
before they can be persisted.

This is better behavior as it prevents invalid jobs from being created
in the first place, rather than allowing them to be created and then
failing during execution.

ferruvich commented on apps/provisioning/pkg/jobs/validator.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	// Validate action-specific options
+	switch job.Spec.Action {
```
Related to the `validActions` check above, I guess we could have the same identical resource if we have a `default` clause here and append an error on the list in case the `switch` falls there?

ferruvich reviewed (approved):
LGTM! Left few comments

MissingRoberto committed:
Simplify action validation using slices.Contains

Replaced manual loop with slices.Contains for cleaner, more idiomatic Go code.
This reduces code complexity while maintaining the same validation logic.

- Added import for 'slices' package
- Replaced 8-line loop with 1-line slices.Contains call
- All unit tests pass

MissingRoberto committed:
Refactor job action validation in ValidateJob function

Removed the hardcoded valid actions array and simplified the validation logic. The function now directly appends an error for invalid actions, improving code clarity and maintainability. This change aligns with the recent updates to job validation, ensuring that invalid job specifications are properly handled.

MissingRoberto commented on apps/provisioning/pkg/jobs/validator.go:
```diff
@@ -0,0 +1,190 @@
+	list := field.ErrorList{}
+
+	// Validate branch name if specified
+	if opts.Branch != "" {
```
Good question! 

Across the codebase, we apply these defaults:
- Empty branch → use the configured branch
- Empty path → use the repository root

These defaults are used by export and other components, so they should remain optional.
